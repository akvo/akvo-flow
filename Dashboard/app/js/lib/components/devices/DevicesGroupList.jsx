import React, { useContext } from 'react';
import DevicesTabContext from './devices-context';
import TABLE_NAMES from './constants';

export default function DevicesGroupList() {
  const {
    strings,
    onSortGroup,
    sortProperties,
    selectGroup,
    devicesGroup,
    tableHeaderClass,
    selectedDeviceGroupIds,
    setCurrentTable,
    onDeleteGroup,
    addNewGroup,
    toggleEditButton,
    selectedEditGroupId,
    renameGroup,
  } = useContext(DevicesTabContext);

  return (
    <>
      <div className="deviceControls">
        <button
          type="button"
          className="btnOutline"
          onClick={() => setCurrentTable(TABLE_NAMES.DEVICES)}
        >
          {strings.navText.manageDevices}
        </button>
        <nav className="dataTabMenu">
          <ul>
            <li>
              <button type="button" onClick={addNewGroup}>
                {strings.navText.newGroup}
              </button>
            </li>
          </ul>
        </nav>
      </div>

      {/* TABLE CONTENTS */}
      <table className="dataTable" id="surveyDataTable">
        <thead>
          <tr>
            <th className="noArrows" />
            <th
              id="device_table_header"
              className={sortProperties.column == 'code' ? tableHeaderClass() : ''}
              onClick={() => onSortGroup('code')}
              onKeyDown={() => onSortGroup('code')}
            >
              {strings.deviceGroup}
            </th>
            <th id="device_table_header" className="noArrows">
              {strings.action}
            </th>
          </tr>
        </thead>
        <tbody>
          {devicesGroup.map(group => {
            const selectedToEdit = selectedEditGroupId === group.keyId;
            return (
              <tr key={group.keyId}>
                <td className="selection">
                  <input
                    type="checkBox"
                    checked={selectedDeviceGroupIds.includes(group.keyId)}
                    onChange={() => selectGroup(group.keyId)}
                  />
                </td>
                <td className="deviceGroup">
                  <button
                    type="button"
                    name={selectedToEdit ? 'save' : 'edit'}
                    id={group.keyId}
                    onClick={toggleEditButton}
                    className={selectedToEdit ? `saveGroupName` : `editGroupName`}
                  >
                    {selectedToEdit ? `Save group name` : `Edit group name`}
                  </button>
                  <input
                    type="text"
                    id={group.keyId}
                    className="editGroupInput"
                    style={{ display: selectedToEdit ? 'block' : 'none' }}
                    defaultValue={group.code}
                    onChange={e => renameGroup({ id: group.keyId, value: e.target.value })}
                  />
                  <span style={{ display: selectedToEdit ? 'none' : 'block' }}>{group.code}</span>
                </td>
                <td>
                  <div
                    onClick={() => onDeleteGroup(group.keyId)}
                    onKeyDown={() => onDeleteGroup(group.keyId)}
                  >
                    {strings.delete}
                  </div>
                </td>
              </tr>
            );
          })}
        </tbody>
      </table>
    </>
  );
}
