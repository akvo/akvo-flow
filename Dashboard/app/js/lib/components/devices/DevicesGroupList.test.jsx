import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import DevicesTabContext from './devices-context';
import DevicesGroupList from './DevicesGroupList';

// Main Tests
describe('DevicesGroupList Tests', () => {
  afterEach(cleanup);

  const devicesGroup = [
    {
      keyId: 1689003,
      description: null,
      name: null,
      code: 'caetie',
      createdDateTime: 1363622757899,
      lastUpdateDateTime: 1363622757899,
      deviceList: null,
      displayName: 'caetie',
    },
  ];

  const strings = {
    deviceGroup: 'Device group',
    delete: 'Delete',
    action: 'Action',
    navText: {
      manageDevices: 'manage device',
      newGroup: 'New group',
    },
  };

  const sortProperties = {
    column: 'code',
  };

  const tableHeaderClass = () => 'sorting_asc';

  const selectedDeviceGroupIds = [];

  it('+++ renders <snapshot>', () => {
    const wrapper = render(
      <DevicesTabContext.Provider
        value={{ devicesGroup, strings, sortProperties, tableHeaderClass, selectedDeviceGroupIds }}
      >
        <DevicesGroupList strings={strings} />
      </DevicesTabContext.Provider>
    );

    const addNewGroupButton = wrapper.getByText('New group', { selector: 'button' });
    fireEvent.click(addNewGroupButton);

    expect(wrapper.container).toMatchSnapshot();
  });
});
