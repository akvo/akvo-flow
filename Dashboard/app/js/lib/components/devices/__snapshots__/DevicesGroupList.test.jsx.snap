// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DevicesGroupList Tests +++ renders <snapshot> 1`] = `
<div>
  <div
    class="deviceControls"
  >
    <button
      class="btnOutline"
      type="button"
    >
      manage device
    </button>
    <nav
      class="dataTabMenu"
    >
      <ul>
        <li>
          <button
            type="button"
          >
            New group
          </button>
        </li>
      </ul>
    </nav>
  </div>
  <table
    class="dataTable"
    id="surveyDataTable"
  >
    <thead>
      <tr>
        <th
          class="sorting_asc"
          id="device_table_header"
        >
          Device group
        </th>
        <th
          class="noArrows"
          id="device_table_header"
        >
          Action
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          class="deviceGroup"
        >
          <button
            class="editGroupName"
            id="1689003"
            name="edit"
            type="button"
          >
            Edit group name
          </button>
          <input
            class="editGroupInput"
            id="1689003"
            style="display: none;"
            type="text"
            value="caetie"
          />
          <span
            style="display: block;"
          >
            caetie
          </span>
        </td>
        <td>
          <div
            id="1689003"
          >
            Delete
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
